src/Foundatio/Messaging/MessageBusBase.cs

```diff
@@ -80,6 +80,18 @@ public abstract class MessageBusBase<TOptions> : IMessageBus, IDisposable where
                     }
 
                     return handler((T)message, cancellationToken);
+                },
+                MessageFilter = (message) => {
```

Comment:
Wondering if we should just keep this null by default. I kind of like having the type filtering outside where it was before as it makes it so this func is kind of filtered before we even get to it and it saves serialization overhead.

Reply:
when we introduce the second interface method we can do it like that

```suggestion
                 public async Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, Func<T, bool> filter, CancellationToken cancellationToken = default) where T : class {
            if (_logger.IsEnabled(LogLevel.Trace))
                _logger.LogTrace("Adding subscriber for {MessageType}.", typeof(T).FullName);
            await EnsureTopicSubscriptionAsync(cancellationToken).AnyContext();
            await SubscribeImplAsync(handler, cancellationToken, filter).AnyContext();
        }

        public async Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, CancellationToken cancellationToken = default) where T : class {
            if (_logger.IsEnabled(LogLevel.Trace))
                _logger.LogTrace("Adding subscriber for {MessageType}.", typeof(T).FullName);
            await EnsureTopicSubscriptionAsync(cancellationToken).AnyContext();
            await SubscribeImplAsync(handler, cancellationToken, msg => true).AnyContext();
        }
        
        ,
                MessageFilter = (message) => {
               
                    if (!(message is T)) {
                        if (_logger.IsEnabled(LogLevel.Trace))
                            _logger.LogTrace("Unable to call subscriber messagefilter: {MessageType} cannot be safely casted to {SubscriberType}", message.GetType(), typeof(T));
                        return true;
                    }

                    return filter((T)message);
                }
        
```

maybe we can integrate the filter into the subscriber action

Reply:
I'd rather have one interface and method. I wonder if we can put this on the options object. There is a unified messaging branch that I believe has this options class, perhaps we can introduce it here and build off it in the other branch? @ejsmith